resolving a method on an embedded binary interface

### The following program `sample.go` triggers an unexpected result

```go
package main

import (
    "fmt"
    "io"
    "net"
)

type wrappedConn struct {
    net.Conn
}

func main() {
    listener, err := net.Listen("tcp", ":49153")
    if err != nil {
        panic(err)
    }
    go func() {
        _, err := listener.Accept()
        if err != nil {
            panic(err)
        }
    }()

    dialer := &net.Dialer{
        LocalAddr: &net.TCPAddr{
            IP:   net.ParseIP("127.0.0.1"),
            Port: 0,
        },
    }

    conn, err := dialer.Dial("tcp", "127.0.0.1:49153")
    if err != nil {
        panic(err)
    }

    t := &wrappedConn{conn}
    var w io.Writer = t
    fmt.Println(w.Write != nil)
}
```


### Expected result

```console
$ go run ./sample.go
true
```


### Got

```console
$ yaegi ./sample.go
./sample.go:14:19: panic
run: ./sample.go:35:20: method not found: Write
goroutine 1 [running]:
runtime/debug.Stack()
	/Users/marc/sdk/go1.17.5/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).Execute.func1(0x14000251ae0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/program.go:131 +0x8c
panic({0x1019ae2a0, 0x140001a93e0})
	/Users/marc/sdk/go1.17.5/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x140001e4c60, 0x140004e0240, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/run.go:194 +0x164
panic({0x1019ae2a0, 0x140001a93e0})
	/Users/marc/sdk/go1.17.5/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.genInterfaceWrapper.func1(0x140001e4c60)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/run.go:1096 +0xa6c
github.com/traefik/yaegi/interp.assign.func3(0x140001e4c60)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/run.go:695 +0x154
github.com/traefik/yaegi/interp.runCfg(0x140004e0240, 0x140001e4c60, 0x14000461680, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/run.go:202 +0x268
github.com/traefik/yaegi/interp.(*Interpreter).run(0x1400020fb00, 0x14000461680, 0x140001e4a50)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/run.go:121 +0x3d4
github.com/traefik/yaegi/interp.(*Interpreter).Execute(0x1400020fb00, 0x140004f30b0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/program.go:157 +0x1e0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x1400020fb00, {0x140004d66c0, 0x22c}, {0x16efaf778, 0x18}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/interp.go:615 +0x98
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x1400020fb00, {0x16efaf778, 0x18})
	/Users/marc/go/src/github.com/traefik/yaegi-radix/interp/interp.go:508 +0xe8
main.runFile(0x1400020fb00, {0x16efaf778, 0x18}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi-radix/cmd/yaegi/run.go:153 +0xb0
main.run({0x140001c0010, 0x1, 0x1})
	/Users/marc/go/src/github.com/traefik/yaegi-radix/cmd/yaegi/run.go:116 +0xe14
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi-radix/cmd/yaegi/yaegi.go:144 +0x384
```


### Yaegi Version

v0.11.1

### Additional Notes

This issue was found when interpreting `gitHub.com/mediocregopher/radix/v4` in yaegi.

--- a/_test/issue-1330.go
+++ b/_test/issue-1330.go
@@ -1,0 +1,43 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+)
+
+type wrappedConn struct {
+	net.Conn
+}
+
+func main() {
+	listener, err := net.Listen("tcp", "127.0.0.1:49153")
+	if err != nil {
+		panic(err)
+	}
+	go func() {
+		_, err := listener.Accept()
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	dialer := &net.Dialer{
+		LocalAddr: &net.TCPAddr{
+			IP:   net.ParseIP("127.0.0.1"),
+			Port: 0,
+		},
+	}
+
+	conn, err := dialer.Dial("tcp", "127.0.0.1:49153")
+	if err != nil {
+		panic(err)
+	}
+
+	t := &wrappedConn{conn}
+	var w io.Writer = t
+	fmt.Println(w.Write != nil)
+}
+
+// Output:
+// true

--- a/_test/issue-1330.go
+++ b/_test/issue-1330.go
@@ -11,16 +11,10 @@ type wrappedConn struct {
 }
 
 func main() {
-	listener, err := net.Listen("tcp", "127.0.0.1:49153")
+	_, err := net.Listen("tcp", "127.0.0.1:49153")
 	if err != nil {
 		panic(err)
 	}
-	go func() {
-		_, err := listener.Accept()
-		if err != nil {
-			panic(err)
-		}
-	}()
 
 	dialer := &net.Dialer{
 		LocalAddr: &net.TCPAddr{
@@ -33,11 +27,16 @@ func main() {
 	if err != nil {
 		panic(err)
 	}
+	defer conn.Close()
 
 	t := &wrappedConn{conn}
 	var w io.Writer = t
-	fmt.Println(w.Write != nil)
+	if n, err := w.Write([]byte("hello")); err != nil {
+		fmt.Println(err)
+	} else {
+		fmt.Println(n)
+	}
 }
 
 // Output:
-// true
+// 5

--- a/interp/interp_consistent_test.go
+++ b/interp/interp_consistent_test.go
@@ -52,6 +52,7 @@ func TestInterpConsistencyBuild(t *testing.T) {
 			file.Name() == "io0.go" || // use random number
 			file.Name() == "issue-1093.go" || // expect error
 			file.Name() == "issue-1276.go" || // expect error
+			file.Name() == "issue-1330.go" || // expect error
 			file.Name() == "op1.go" || // expect error
 			file.Name() == "op7.go" || // expect error
 			file.Name() == "op9.go" || // expect error

--- a/interp/run.go
+++ b/interp/run.go
@@ -1126,12 +1126,35 @@ func methodByName(value reflect.Value, name string, index []int) (v reflect.Valu
 	}
 	for value.Kind() == reflect.Ptr {
 		value = value.Elem()
-		o := value.FieldByIndex(index)
-		if v = o.MethodByName(name); v.IsValid() {
+		if checkFieldIndex(value.Type(), index) {
+			value = value.FieldByIndex(index)
+		}
+		if v = value.MethodByName(name); v.IsValid() {
 			return
 		}
 	}
 	return
+}
+
+func checkFieldIndex(typ reflect.Type, index []int) bool {
+	if len(index) == 0 {
+		return false
+	}
+	t := typ
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	if t.Kind() != reflect.Struct {
+		return false
+	}
+	i := index[0]
+	if i >= t.NumField() {
+		return false
+	}
+	if len(index) > 1 {
+		return checkFieldIndex(t.Field(i).Type, index[1:])
+	}
+	return true
 }
 
 func call(n *node) {

--- a/interp/run.go
+++ b/interp/run.go
@@ -1086,7 +1086,7 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 		for i, m := range methods {
 			if m == nil {
 				// First direct method lookup on field.
-				if r := methodByName(v, names[i]); r.IsValid() {
+				if r := methodByName(v, names[i], indexes[i]); r.IsValid() {
 					w.Field(i + 1).Set(r)
 					continue
 				}
@@ -1115,7 +1115,7 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 // The search is extended on valueInterface wrapper if present.
 // If valid, the returned value is a method function with the receiver already set
 // (no need to pass it at call).
-func methodByName(value reflect.Value, name string) (v reflect.Value) {
+func methodByName(value reflect.Value, name string, index []int) (v reflect.Value) {
 	if vi, ok := value.Interface().(valueInterface); ok {
 		if v = getConcreteValue(vi.value).MethodByName(name); v.IsValid() {
 			return
@@ -1126,7 +1126,8 @@ func methodByName(value reflect.Value, name string) (v reflect.Value) {
 	}
 	for value.Kind() == reflect.Ptr {
 		value = value.Elem()
-		if v = value.MethodByName(name); v.IsValid() {
+		o := value.FieldByIndex(index)
+		if v = o.MethodByName(name); v.IsValid() {
 			return
 		}
 	}

--- a/interp/run.go
+++ b/interp/run.go
@@ -1111,15 +1111,26 @@ func genInterfaceWrapper(n *node, typ reflect.Type) func(*frame) reflect.Value {
 	}
 }
 
-// methodByName return the method corresponding to name on value, or nil if not found.
+// methodByName returns the method corresponding to name on value, or nil if not found.
 // The search is extended on valueInterface wrapper if present.
-func methodByName(value reflect.Value, name string) reflect.Value {
+// If valid, the returned value is a method function with the receiver already set
+// (no need to pass it at call).
+func methodByName(value reflect.Value, name string) (v reflect.Value) {
 	if vi, ok := value.Interface().(valueInterface); ok {
-		if v := getConcreteValue(vi.value).MethodByName(name); v.IsValid() {
-			return v
+		if v = getConcreteValue(vi.value).MethodByName(name); v.IsValid() {
+			return
 		}
 	}
-	return value.MethodByName(name)
+	if v = value.MethodByName(name); v.IsValid() {
+		return
+	}
+	for value.Kind() == reflect.Ptr {
+		value = value.Elem()
+		if v = value.MethodByName(name); v.IsValid() {
+			return
+		}
+	}
+	return
 }
 
 func call(n *node) {
@@ -1610,11 +1621,16 @@ func callBin(n *node) {
 				}
 				out := callFn(value(f), in)
 				for i := 0; i < len(out); i++ {
-					if out[i].Kind() == reflect.Func {
-						getFrame(f, n.level).data[n.findex+i] = out[i]
-					} else {
-						getFrame(f, n.level).data[n.findex+i].Set(out[i])
+					r := out[i]
+					if r.Kind() == reflect.Func {
+						getFrame(f, n.level).data[n.findex+i] = r
+						continue
 					}
+					dest := getFrame(f, n.level).data[n.findex+i]
+					if _, ok := dest.Interface().(valueInterface); ok {
+						r = reflect.ValueOf(valueInterface{value: r})
+					}
+					dest.Set(r)
 				}
 				return tnext
 			}

